dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+//  * For new capacity of zero, free() the buffer and set the pointer to NULL
+//  * When reducing length, call destructor for any items which are being removed
+void buffer_alloc(struct buffer *inst, size_t new_capacity)
+{  
+    if (new_capacity == 0){
+        buffer_destroy(inst);
+    } else{
+        // realloc will free the necessary memory
+        try_realloc(inst, new_capacity);
+    }
+}
+
+// Grow the buffer if necessary, to ensure the requested minimum capacity
+void buffer_reserve(struct buffer *inst, size_t min_capacity)
+{
+    if(inst->capacity < min_capacity) {
+        size_t previous_capacity = inst->capacity;
+        inst->buffer = realloc_zero(inst->buffer, 
+                       previous_capacity * sizeof(void*), 
+                       min_capacity * sizeof(void*));
+        
+        for (size_t i = inst->capacity; i < min_capacity; ++i) {
+            void *succeds =  realloc(inst->buffer[i], inst->item_size);
+            if (succeds){
+                inst->buffer[i] = succeds;
+            } else {
+                exit(0);
+            }
+        }
+        inst->capacity = min_capacity;
+    }
+}
+
+// If realloc needs to increase size, safely increase and 
+// memset the new data to zero;
+void* realloc_zero(void* old_buffer, size_t old_size, size_t new_size) 
+{
+    void* new_buffer = realloc(old_buffer, new_size);
+    if (new_size > old_size && new_buffer) {
+        size_t diff = new_size - old_size;
+        void* start = new_buffer + old_size;
+        memset(start, 0, diff);
+    } else {
+        exit(0);
+    }
+    return new_buffer;
+}
+
+// Change length of buffer
+//  * If growing, grow capacity if required and zero-fill new items
+//  * If shrinking, call destructor for any items which are being removed
+void buffer_resize(struct buffer *inst, size_t new_length)
+{   
+    // growing
+    if (inst->capacity < new_length) {
+        buffer_reserve(inst, new_length);
+    // shrinking
+    } else if (inst->capacity > new_length) {
+        for (size_t i = new_length; i < inst->capacity; ++i) {
+            ((destr)(inst->destructor))(inst->buffer[i]);
+        }
+        void *succeds = realloc(inst->buffer, new_length *sizeof(void*));
+        if (succeds) {
+            inst->buffer = succeds;
+        } else {
+            exit(0);
+        }
+    }
+    inst->capacity = new_length;
+    if (inst->length > new_length) {
+        inst->length = inst->capacity;
+    }
+}
+
+// Append to buffer (if no space, grow buffer by alloc_by items)
+void buffer_push_back(struct buffer *inst, void *item) 
+{   
+    size_t length = buffer_size(inst);
+    ++inst->length;
+    if (length < inst->capacity) {
+        void *new_item = buffer_at(inst, length);
+        if (item) {
+            memcpy(new_item, item, inst->item_size);
+        } else {
+            --inst->length;
+        }
+    } else {
+        buffer_reserve(inst, inst->alloc_by + length);
```

Comment:
When auto-growing the buffer in push_back, we don't actually seem to push the item after growing.